# Bellabeat Case Study — Process Phase

## Activity Dataset Processing

## Load datasets
Load the two daily activity files covering the two collection periods, along with the sleep dataset, into separate Pandas DataFrames.

In [190]:
import pandas as pd

activity_early = pd.read_csv(
    "../data/working data/dailyActivity_2016-03-12_to_2016-04-11.csv"
)

activity_late = pd.read_csv(
    "../data/working data/dailyActivity_2016-04-12_to_2016-05-12.csv"
)

## Looking at datasets
Inspected the first five rows of both activity datasets to confirm that the files loaded correctly and to review their structure and values.

In [191]:
display(activity_early.head())
display(activity_late.head())

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,3/25/2016,11004,7.11,7.11,0.0,2.57,0.46,4.07,0.0,33,12,205,804,1819
1,1503960366,3/26/2016,17609,11.55,11.55,0.0,6.92,0.73,3.91,0.0,89,17,274,588,2154
2,1503960366,3/27/2016,12736,8.53,8.53,0.0,4.66,0.16,3.71,0.0,56,5,268,605,1944
3,1503960366,3/28/2016,13231,8.93,8.93,0.0,3.19,0.79,4.95,0.0,39,20,224,1080,1932
4,1503960366,3/29/2016,12041,7.85,7.85,0.0,2.16,1.09,4.61,0.0,28,28,243,763,1886


,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,4/12/2016,13162,8.50,8.50,0.0,1.88,0.55,6.06,0.0,25,13,328,728,1985
1,1503960366,4/13/2016,10735,6.97,6.97,0.0,1.57,0.69,4.71,0.0,21,19,217,776,1797
2,1503960366,4/14/2016,10460,6.74,6.74,0.0,2.44,0.40,3.91,0.0,30,11,181,1218,1776
3,1503960366,4/15/2016,9762,6.28,6.28,0.0,2.14,1.26,2.83,0.0,29,34,209,726,1745
4,1503960366,4/16/2016,12669,8.16,8.16,0.0,2.71,0.41,5.04,0.0,36,10,221,773,1863


## SHAPE
(rows, columns)
Compared the dimensions of both activity datasets. The early-period dataset contains 457 rows and 15 columns, while the later-period dataset contains 940 rows and 15 columns. Both datasets therefore contain the same number of variables.

In [192]:
print(activity_early.shape)
print(activity_late.shape)

(457, 15)
(940, 15)


## COLUMNS
Reviewed the column names in both datasets to confirm that they contain the same activity variables before combining them.

In [193]:
print(activity_early.columns)
print(activity_late.columns)

Index(['Id', 'ActivityDate', 'TotalSteps', 'TotalDistance', 'TrackerDistance',
       'LoggedActivitiesDistance', 'VeryActiveDistance',
       'ModeratelyActiveDistance', 'LightActiveDistance',
       'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes',
       'LightlyActiveMinutes', 'SedentaryMinutes', 'Calories'],
      dtype='str')
Index(['Id', 'ActivityDate', 'TotalSteps', 'TotalDistance', 'TrackerDistance',
       'LoggedActivitiesDistance', 'VeryActiveDistance',
       'ModeratelyActiveDistance', 'LightActiveDistance',
       'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes',
       'LightlyActiveMinutes', 'SedentaryMinutes', 'Calories'],
      dtype='str')


## CHECK IF COLUMNS MATCH EXACTLY
Confirmed that both datasets contain identical column names in the same order, indicating that their schemas are compatible for concatenation.

In [194]:
activity_early.columns.equals(activity_late.columns)

True

## DATES
Converted ActivityDate to a datetime data type to support accurate date comparisons and chronological analysis. The minimum and maximum dates were then reviewed to establish the collection period covered by each dataset.

In [195]:
activity_early['ActivityDate'] = pd.to_datetime(activity_early['ActivityDate'])
activity_late['ActivityDate'] = pd.to_datetime(activity_late['ActivityDate'])

print(activity_early['ActivityDate'].min(), activity_early['ActivityDate'].max())
print(activity_late['ActivityDate'].min(), activity_late['ActivityDate'].max())

2016-03-12 00:00:00 2016-04-12 00:00:00
2016-04-12 00:00:00 2016-05-12 00:00:00


## OVERLAP
Compared the dates present in both datasets to determine whether the collection periods overlap. An overlap was identified on 12 April 2016, requiring further investigation before combining the files.

In [196]:
overlap = set(activity_early['ActivityDate']).intersection(
    set(activity_late['ActivityDate'])
)

print(overlap)

{Timestamp('2016-04-12 00:00:00')}


## INSPECT OVERLAP BEFORE COMBINING
Examined records from 12 April 2016 in both datasets to determine how many observations were present and whether the same participants appeared in both files.

In [197]:
display(
    activity_early[activity_early['ActivityDate'] == '2016-04-12']
)

display(
    activity_late[activity_late['ActivityDate'] == '2016-04-12']
)

early_apr12 = activity_early[
    activity_early['ActivityDate'] == '2016-04-12'
]

late_apr12 = activity_late[
    activity_late['ActivityDate'] == '2016-04-12'
]

print("Early rows:", len(early_apr12))
print("Late rows:", len(late_apr12))

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
18,1503960366,2016-04-12,224,0.14,0.14,0.000000,0.00,0.00,0.13,0.0,0,0,9,32,50
37,1624580081,2016-04-12,6627,4.31,4.31,0.000000,0.00,0.00,4.31,0.0,0,0,89,604,706
59,1844505072,2016-04-12,0,0.00,0.00,0.000000,0.00,0.00,0.00,0.0,0,0,0,426,399
71,1927972279,2016-04-12,24,0.02,0.02,0.000000,0.00,0.00,0.02,0.0,0,0,3,161,942
83,2022484408,2016-04-12,6717,4.72,4.72,0.000000,3.23,0.22,1.27,0.0,41,5,67,551,1140
95,2026352035,2016-04-12,1019,0.63,0.63,0.000000,0.00,0.00,0.63,0.0,0,0,64,223,600
107,2320127002,2016-04-12,2098,1.41,1.41,0.000000,0.00,0.00,1.41,0.0,0,0,69,617,790
122,2347167796,2016-04-12,0,0.00,0.00,0.000000,0.00,0.00,0.00,0.0,0,0,0,425,399
134,2873212765,2016-04-12,0,0.00,0.00,0.000000,0.00,0.00,0.00,0.0,0,0,0,1440,0
164,3977333714,2016-04-12,759,0.57,0.57,0.000000,0.00,0.00,0.57,0.0,0,0,17,187,182


,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,2016-04-12,13162,8.50,8.50,0.000000,1.88,0.55,6.06,0.00,25,13,328,728,1985
31,1624580081,2016-04-12,8163,5.31,5.31,0.000000,0.00,0.00,5.31,0.00,0,0,146,1294,1432
62,1644430081,2016-04-12,10694,7.77,7.77,0.000000,0.14,2.30,5.33,0.00,2,51,256,1131,3199
92,1844505072,2016-04-12,6697,4.43,4.43,0.000000,0.00,0.00,4.43,0.00,0,0,339,1101,2030
123,1927972279,2016-04-12,678,0.47,0.47,0.000000,0.00,0.00,0.47,0.00,0,0,55,734,2220
154,2022484408,2016-04-12,11875,8.34,8.34,0.000000,3.31,0.77,4.26,0.00,42,14,227,1157,2390
185,2026352035,2016-04-12,4414,2.74,2.74,0.000000,0.19,0.35,2.20,0.00,3,8,181,706,1459
216,2320127002,2016-04-12,10725,7.49,7.49,0.000000,1.17,0.31,6.01,0.00,13,9,306,1112,2124
247,2347167796,2016-04-12,10113,6.83,6.83,0.000000,2.00,0.62,4.20,0.00,28,13,320,964,2344
265,2873212765,2016-04-12,8796,5.91,5.91,0.000000,0.11,0.93,4.88,0.00,2,21,356,1061,1982


Early rows: 24
Late rows: 33


## IDENTIFY PARTICIPANTS PRESENT IN BOTH FILES
Compared participant IDs recorded on 12 April 2016 across both datasets to identify users represented in both collection periods.

In [198]:
overlap_ids = set(early_apr12['Id']).intersection(
    set(late_apr12['Id'])
)

print(overlap_ids)

{8053475328, 8378563200, 4558609924, 2320127002, 2026352035, 7007744171, 1503960366, 8877689391, 2347167796, 1927972279, 2022484408, 8792009665, 5553957443, 7086361926, 6962181067, 4057192912, 3977333714, 4702921684, 2873212765, 4020332650, 4445114986, 1844505072, 1624580081, 8253242879}


## CHECK DUPLICATES
After temporarily concatenating the datasets, 24 duplicate Id + ActivityDate combinations were identified. This indicated that some participants had records for 12 April 2016 in both files.

In [199]:
activity_test = pd.concat(
    [activity_early, activity_late],
    ignore_index=True
)

activity_test.duplicated(
    subset=['Id', 'ActivityDate']
).sum()

np.int64(24)

## INSPECT DUPLICATES 
Reviewed the overlapping participant-date records to determine whether they were exact duplicate rows or separate records containing conflicting activity measurements.

In [200]:
duplicates = activity_test[
    activity_test.duplicated(
        subset=['Id', 'ActivityDate'], 
        keep=False
    )
]

duplicates.sort_values(['Id', 'ActivityDate'])

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
18,1503960366,2016-04-12,224,0.14,0.14,0.000000,0.00,0.00,0.13,0.0,0,0,9,32,50
457,1503960366,2016-04-12,13162,8.50,8.50,0.000000,1.88,0.55,6.06,0.0,25,13,328,728,1985
37,1624580081,2016-04-12,6627,4.31,4.31,0.000000,0.00,0.00,4.31,0.0,0,0,89,604,706
488,1624580081,2016-04-12,8163,5.31,5.31,0.000000,0.00,0.00,5.31,0.0,0,0,146,1294,1432
59,1844505072,2016-04-12,0,0.00,0.00,0.000000,0.00,0.00,0.00,0.0,0,0,0,426,399
549,1844505072,2016-04-12,6697,4.43,4.43,0.000000,0.00,0.00,4.43,0.0,0,0,339,1101,2030
71,1927972279,2016-04-12,24,0.02,0.02,0.000000,0.00,0.00,0.02,0.0,0,0,3,161,942
580,1927972279,2016-04-12,678,0.47,0.47,0.000000,0.00,0.00,0.47,0.0,0,0,55,734,2220
83,2022484408,2016-04-12,6717,4.72,4.72,0.000000,3.23,0.22,1.27,0.0,41,5,67,551,1140
611,2022484408,2016-04-12,11875,8.34,8.34,0.000000,3.31,0.77,4.26,0.0,42,14,227,1157,2390


## CHECK FOR EXACT DUPLICATE ROWS
No fully identical rows were found. This confirmed that the overlapping records represented the same participant and date but contained different activity measurements.

In [201]:
activity_test.duplicated().sum()

np.int64(0)

## CONFIRM DATE RANGES
The early dataset covers 12 March–12 April 2016, while the later dataset covers 12 April–12 May 2016, confirming that the collection periods overlap on 12 April 2016.

In [202]:
print(activity_early['ActivityDate'].min(), activity_early['ActivityDate'].max())
print(activity_late['ActivityDate'].min(), activity_late['ActivityDate'].max())

2016-03-12 00:00:00 2016-04-12 00:00:00
2016-04-12 00:00:00 2016-05-12 00:00:00


## REMOVE THE OVERLAP
Both datasets contain records for 12 April 2016. Because this date represents the start of the second collection period, and the later dataset contains records for all 33 participants compared with only 24 in the earlier dataset, the 12 April observations from the later dataset were retained and those from the earlier dataset were excluded. A cleaned version of the early dataset was created before concatenation.

In [203]:
activity_early_clean = activity_early[
    activity_early['ActivityDate'] < '2016-04-12'
]

## COMBINE THE TWO DATASETS
Concatenated the cleaned early-period dataset with the later-period dataset to create a single activity DataFrame. The index was reset to provide a continuous row index across the combined data.

In [204]:
activity_all = pd.concat(
    [activity_early_clean, activity_late],
    ignore_index=True
)

## VERIFY OVERLAP IS GONE
Rechecked the combined data for duplicate Id + ActivityDate combinations. The result of 0 confirmed that each participant-date combination is now unique.

In [205]:
activity_all.duplicated(
    subset=['Id', 'ActivityDate']
).sum()

np.int64(0)

## CHECK FINAL SHAPE

In [206]:
activity_all.shape

(1373, 15)

## CHECK MISSING VALUES
No missing values were identified across the 15 columns of the combined activity dataset.

In [207]:
activity_all.isnull().sum()

Id                          0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        0
SedentaryMinutes            0
Calories                    0
dtype: int64

## CHECK DATA TYPES
Reviewed the data types of all variables to confirm that numeric activity measures were stored appropriately and that ActivityDate had been successfully converted to a datetime field.

In [208]:
activity_all.info()

<class 'pandas.DataFrame'>
RangeIndex: 1373 entries, 0 to 1372
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Id                        1373 non-null   int64         
 1   ActivityDate              1373 non-null   datetime64[us]
 2   TotalSteps                1373 non-null   int64         
 3   TotalDistance             1373 non-null   float64       
 4   TrackerDistance           1373 non-null   float64       
 5   LoggedActivitiesDistance  1373 non-null   float64       
 6   VeryActiveDistance        1373 non-null   float64       
 7   ModeratelyActiveDistance  1373 non-null   float64       
 8   LightActiveDistance       1373 non-null   float64       
 9   SedentaryActiveDistance   1373 non-null   float64       
 10  VeryActiveMinutes         1373 non-null   int64         
 11  FairlyActiveMinutes       1373 non-null   int64         
 12  LightlyActiveMinutes      1373 

## CHANGE DATE FORMAT

In [209]:
activity_all['ActivityDate_UK'] = activity_all['ActivityDate'].dt.strftime('%d/%m/%Y')

activity_all[['ActivityDate', 'ActivityDate_UK']].head()

,ActivityDate,ActivityDate_UK
0,2016-03-25,25/03/2016
1,2016-03-26,26/03/2016
2,2016-03-27,27/03/2016
3,2016-03-28,28/03/2016
4,2016-03-29,29/03/2016


## CHECK EXACT DUPLICATES AGAIN

In [210]:
activity_all.duplicated().sum()

np.int64(0)

## SANITY CHECK
Reviewed summary statistics to identify potentially invalid or implausible values, including negative steps, calories, distances, or activity minutes, as well as unusually large observations that may require further investigation.

In [211]:
activity_all.describe()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
count,1.373000e+03,1373,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000
mean,4.782326e+09,2016-04-19 04:24:17.829570,7377.375091,5.289031,5.261930,0.128591,1.413146,0.546052,3.238128,0.001733,19.871085,13.603787,188.096140,1001.346686,2294.812090
min,1.503960e+09,2016-03-12 00:00:00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2.320127e+09,2016-04-08 00:00:00,3321.000000,2.280000,2.270000,0.000000,0.000000,0.000000,1.730000,0.000000,0.000000,0.000000,117.000000,734.000000,1820.000000
50%,4.445115e+09,2016-04-19 00:00:00,7142.000000,5.030000,5.020000,0.000000,0.120000,0.200000,3.280000,0.000000,2.000000,6.000000,196.000000,1062.000000,2129.000000
75%,6.962181e+09,2016-04-30 00:00:00,10645.000000,7.570000,7.570000,0.000000,1.860000,0.780000,4.710000,0.000000,30.000000,18.000000,263.000000,1246.000000,2781.000000
max,8.877689e+09,2016-05-12 00:00:00,36019.000000,28.030001,28.030001,6.727057,21.920000,6.480000,12.510000,0.110000,210.000000,660.000000,720.000000,1440.000000,4900.000000
std,2.381544e+09,NaN,5198.133424,3.987341,3.973439,0.695630,2.622729,0.872581,2.104300,0.007800,31.817631,26.581555,113.095058,304.168197,725.526989


In [212]:
(activity_all['TotalSteps'] < 0).sum()

np.int64(0)

## Sleep Dataset Processing

## LOAD AND LOOK AT DATASET
Inspected the first five rows to confirm successful import and review the available sleep variables.

In [213]:
sleep = pd.read_csv("../data/working data/sleepDay_merged.csv")

sleep.head()

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
0,1503960366,4/12/2016 12:00:00 AM,1,327,346
1,1503960366,4/13/2016 12:00:00 AM,2,384,407
2,1503960366,4/15/2016 12:00:00 AM,1,412,442
3,1503960366,4/16/2016 12:00:00 AM,2,340,367
4,1503960366,4/17/2016 12:00:00 AM,1,700,712


## SHAPE AND INFO
Compared the dimensions of both activity datasets. The dataset contains 413 rows and 5 columns.

In [214]:
sleep.shape

(413, 5)

In [215]:
sleep.info()

<class 'pandas.DataFrame'>
RangeIndex: 413 entries, 0 to 412
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   Id                  413 non-null    int64
 1   SleepDay            413 non-null    str  
 2   TotalSleepRecords   413 non-null    int64
 3   TotalMinutesAsleep  413 non-null    int64
 4   TotalTimeInBed      413 non-null    int64
dtypes: int64(4), str(1)
memory usage: 16.3 KB


## COVERT TEXT
Converts the SleepDay column from plain text into a proper datetime type

In [216]:
sleep['SleepDay'] = pd.to_datetime(sleep['SleepDay'])

C:\Users\btari\AppData\Local\Temp\ipykernel_11028\1544153808.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  sleep['SleepDay'] = pd.to_datetime(sleep['SleepDay'])


## MISSING VALUES ?
No missing values were identified across the 5 columns of the combined activity dataset.

In [217]:
sleep.isnull().sum()

Id                    0
SleepDay              0
TotalSleepRecords     0
TotalMinutesAsleep    0
TotalTimeInBed        0
dtype: int64

## EXACT DUPLICATE CHECK 
Check for rows that are identical across every column. Three duplicate records were identified and inspected before removal.

In [218]:
sleep.duplicated().sum()

np.int64(3)

## REMOVE DUPLICATES AND VALIDATE
Exact duplicates found: Three duplicated sleep records were identified. Each duplicated pair contained identical values across all columns, so the duplicate copies were removed before analysis.

In [219]:
sleep[sleep.duplicated(keep=False)]

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
160,4388161847,2016-05-05,1,471,495
161,4388161847,2016-05-05,1,471,495
222,4702921684,2016-05-07,1,520,543
223,4702921684,2016-05-07,1,520,543
379,8378563200,2016-04-25,1,388,402
380,8378563200,2016-04-25,1,388,402


In [220]:
sleep_clean = sleep.drop_duplicates().reset_index(drop=True)

In [221]:
sleep_clean.duplicated().sum()

np.int64(0)

## ID AND DATE DUPLICATE CHECK
Check to see if the same participant has more than one record on the same day? It is expected to have one participant and one sleep day equate to one sleep record

In [222]:
sleep_clean.duplicated(subset=['Id', 'SleepDay']).sum()

np.int64(0)

## SANITY CHECK - DESCRIBE
Check if theres any impossible or suspicious values

In [223]:
sleep_clean.describe()

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
count,4.100000e+02,410,410.000000,410.000000,410.000000
mean,4.994963e+09,2016-04-26 11:38:55.609756,1.119512,419.173171,458.482927
min,1.503960e+09,2016-04-12 00:00:00,1.000000,58.000000,61.000000
25%,3.977334e+09,2016-04-19 00:00:00,1.000000,361.000000,403.750000
50%,4.702922e+09,2016-04-27 00:00:00,1.000000,432.500000,463.000000
75%,6.962181e+09,2016-05-04 00:00:00,1.000000,490.000000,526.000000
max,8.792010e+09,2016-05-12 00:00:00,3.000000,796.000000,961.000000
std,2.060863e+09,NaN,0.346636,118.635918,127.455140


## VALIDITY CHECK - IMPOSSIBLE VALUES
Check if any sleep duration was negative whihc would be impossible

In [224]:
(sleep_clean['TotalMinutesAsleep']<0).sum()

np.int64(0)

In [225]:
(sleep_clean['TotalTimeInBed'] < 0).sum()

np.int64(0)

## SANITY CHECK 
Reviewed summary statistics to identify potentially invalid or implausible values, including if minutes asleep is greater than total time in bed

In [226]:
sleep_clean[['TotalMinutesAsleep', 'TotalTimeInBed']]. describe()

,TotalMinutesAsleep,TotalTimeInBed
count,410.000000,410.000000
mean,419.173171,458.482927
std,118.635918,127.455140
min,58.000000,61.000000
25%,361.000000,403.750000
50%,432.500000,463.000000
75%,490.000000,526.000000
max,796.000000,961.000000


In [227]:
(sleep_clean['TotalMinutesAsleep'] > sleep_clean['TotalTimeInBed']).sum()

np.int64(0)

## DATA FORMAT
Change the Data format to match UK format for better reading comprehendsion and ease of understanding

In [228]:
sleep_clean['SleepDay_UK'] = sleep_clean['SleepDay'].dt.strftime('%d/%m/%Y')

In [229]:
sleep_clean[['SleepDay', 'SleepDay_UK']].head()

,SleepDay,SleepDay_UK
0,2016-04-12,12/04/2016
1,2016-04-13,13/04/2016
2,2016-04-15,15/04/2016
3,2016-04-16,16/04/2016
4,2016-04-17,17/04/2016


# Bellabeat Case Study — Analyze Phase

## INCREASING READABILITY
Changing the column format from minutes to hours for better comprehension of the values

In [230]:
activity_all['SedentaryHours'] = activity_all['SedentaryMinutes'] / 60

activity_all['VeryActiveHours'] = activity_all['VeryActiveMinutes'] / 60
activity_all['FairlyActiveHours'] = activity_all['FairlyActiveMinutes'] / 60
activity_all['LightlyActiveHours'] = activity_all['LightlyActiveMinutes'] / 60

In [231]:
sleep_clean['HoursAsleep'] = sleep_clean['TotalMinutesAsleep'] / 60
sleep_clean['HoursInBed'] = sleep_clean['TotalTimeInBed'] / 60

## OVERALL ACTIVE SUMMARY
Summary statistics are calculated for key activity measures to understand the typical values, variation, and overall range within the dataset.

In [232]:
activity_summary = activity_all[
    [
        'TotalSteps',
        'VeryActiveHours',
        'FairlyActiveHours',
        'LightlyActiveHours',
        'SedentaryHours',
        'Calories'
    ]
].describe().round(2)

activity_summary

,TotalSteps,VeryActiveHours,FairlyActiveHours,LightlyActiveHours,SedentaryHours,Calories
count,1373.00,1373.00,1373.00,1373.00,1373.00,1373.00
mean,7377.38,0.33,0.23,3.13,16.69,2294.81
std,5198.13,0.53,0.44,1.88,5.07,725.53
min,0.00,0.00,0.00,0.00,0.00,0.00
25%,3321.00,0.00,0.00,1.95,12.23,1820.00
50%,7142.00,0.03,0.10,3.27,17.70,2129.00
75%,10645.00,0.50,0.30,4.38,20.77,2781.00
max,36019.00,3.50,11.00,12.00,24.00,4900.00


## OVERALL SLEEP SUMMARY
Calculate summary statistics like count, mean, median, min, max, etc.

## TOTAL AWAKE IN BED TIME

In [233]:
sleep_summary = sleep_clean[
    [
        'TotalSleepRecords',
        'HoursAsleep',
        'HoursInBed'
    ]
].describe().round(2)

sleep_summary

,TotalSleepRecords,HoursAsleep,HoursInBed
count,410.00,410.00,410.00
mean,1.12,6.99,7.64
std,0.35,1.98,2.12
min,1.00,0.97,1.02
25%,1.00,6.02,6.73
50%,1.00,7.21,7.72
75%,1.00,8.17,8.77
max,3.00,13.27,16.02


## TOTAL ACTIVE TIME

In [234]:
sleep_clean['TimeAwakeInBedHours'] = (
    sleep_clean['HoursInBed'] - sleep_clean['HoursAsleep']
)

sleep_clean['TimeAwakeInBedHours'].describe().round(2)

count    410.00
mean       0.66
std        0.78
min        0.00
25%        0.28
50%        0.43
75%        0.67
max        6.18
Name: TimeAwakeInBedHours, dtype: float64

## BETTER READING COMPREHENSION
For better reading comprehension, transformed duration format from decimals to words
TotalActiveHours -> numeric decimal (3.69)
TotalActiveTime -> formatted text (3h 41m)

In [235]:
activity_all['TotalActiveMinutes'] = (
    activity_all['VeryActiveMinutes']
    + activity_all['FairlyActiveMinutes']
    + activity_all['LightlyActiveMinutes']
)

activity_all['TotalActiveHours'] = activity_all['TotalActiveMinutes'] / 60

activity_all['TotalActiveHours'].describe().round(2)

count    1373.00
mean        3.69
std         2.13
min         0.00
25%         2.25
50%         4.07
75%         5.25
max        12.00
Name: TotalActiveHours, dtype: float64

In [236]:
def hours_to_hm(hours):
    total_minutes = round(hours * 60)
    h = total_minutes // 60
    m = total_minutes % 60
    return f"{h}h {m}m"

In [237]:
activity_all['TotalActiveTime'] = activity_all['TotalActiveHours'].apply(hours_to_hm)

sleep_clean['SleepDuration'] = sleep_clean['HoursAsleep'].apply(hours_to_hm)
sleep_clean['TimeInBed'] = sleep_clean['HoursInBed'].apply(hours_to_hm)
sleep_clean['AwakeInBedTime'] = sleep_clean['TimeAwakeInBedHours'].apply(hours_to_hm)

In [238]:
activity_all[
    ['TotalActiveHours', 'TotalActiveTime']
].head()

,TotalActiveHours,TotalActiveTime
0,4.166667,4h 10m
1,6.333333,6h 20m
2,5.483333,5h 29m
3,4.716667,4h 43m
4,4.983333,4h 59m


## WEEKDAY PATTERNS - CREATING COLUMNS
Creating weekday columns

In [239]:
activity_all['DayOfWeek'] = activity_all['ActivityDate'].dt.day_name()

sleep_clean['DayOfWeek'] = sleep_clean['SleepDay'].dt.day_name()

## WEEKDAY ACTIVITY ANALYSIS
Calculate average steps, active hours, sedentary hours, and calories by day of week to identify weekly activity patterns.

In [240]:
activity_by_day = activity_all.groupby('DayOfWeek')[
    [
        'TotalSteps',
        'TotalActiveHours',
        'SedentaryHours',
        'Calories'
    ]
].mean().round(2)

activity_by_day

,TotalSteps,TotalActiveHours,SedentaryHours,Calories
DayOfWeek,,,,
Friday,7187.53,3.82,17.00,2325.10
Monday,7541.32,3.65,17.16,2298.40
Saturday,7752.27,3.93,16.15,2325.80
Sunday,6606.73,3.35,16.66,2227.41
Thursday,7268.30,3.54,16.42,2223.75
Tuesday,7718.96,3.83,16.87,2338.88
Wednesday,7547.58,3.72,16.58,2320.76


In [241]:
sleep_by_day = sleep_clean.groupby('DayOfWeek')[
    [
        'HoursAsleep',
        'HoursInBed',
        'TimeAwakeInBedHours'
    ]
].mean().round(2)

sleep_by_day

,HoursAsleep,HoursInBed,TimeAwakeInBedHours
DayOfWeek,,,
Friday,6.76,7.42,0.66
Monday,6.99,7.62,0.63
Saturday,6.98,7.66,0.68
Sunday,7.55,8.39,0.85
Thursday,6.69,7.25,0.56
Tuesday,6.74,7.39,0.65
Wednesday,7.24,7.83,0.59


## REORDERING WEEKDAYS
Pandas grouped weekday names alphabetically, so I reorder the index to Monday–Sunday for clearer interpretation.

In [242]:
day_order = [
    'Monday',
    'Tuesday',
    'Wednesday',
    'Thursday',
    'Friday',
    'Saturday',
    'Sunday'
]

In [243]:
activity_by_day = activity_by_day.reindex(day_order)
sleep_by_day = sleep_by_day.reindex(day_order)

In [244]:
activity_by_day

,TotalSteps,TotalActiveHours,SedentaryHours,Calories
DayOfWeek,,,,
Monday,7541.32,3.65,17.16,2298.40
Tuesday,7718.96,3.83,16.87,2338.88
Wednesday,7547.58,3.72,16.58,2320.76
Thursday,7268.30,3.54,16.42,2223.75
Friday,7187.53,3.82,17.00,2325.10
Saturday,7752.27,3.93,16.15,2325.80
Sunday,6606.73,3.35,16.66,2227.41


In [245]:
sleep_by_day

,HoursAsleep,HoursInBed,TimeAwakeInBedHours
DayOfWeek,,,
Monday,6.99,7.62,0.63
Tuesday,6.74,7.39,0.65
Wednesday,7.24,7.83,0.59
Thursday,6.69,7.25,0.56
Friday,6.76,7.42,0.66
Saturday,6.98,7.66,0.68
Sunday,7.55,8.39,0.85


## WEEKDAY FINDINGS
Activity levels were highest on Saturday and lowest on Sunday. Sleep duration peaked on Sunday, while Thursday showed the shortest average sleep. This suggests user behaviour varies by day of week, with weekends showing different activity and sleep patterns from weekdays.

## ACTIVITY CORRELATIONS 

In [246]:
activity_corr = activity_all[
    [
        'TotalSteps',
        'Calories',
        'TotalActiveHours',
        'SedentaryHours'
    ]        
].corr().round(2)

activity_corr

,TotalSteps,Calories,TotalActiveHours,SedentaryHours
TotalSteps,1.00,0.58,0.77,-0.36
Calories,0.58,1.00,0.47,-0.13
TotalActiveHours,0.77,0.47,1.00,-0.52
SedentaryHours,-0.36,-0.13,-0.52,1.00


## ACTIVITY CORRELATIONS FINDINGS:
Correlation values range from -1 to +1
close to +1 = strong positive relationship
close to -1 = strong negative relationship
close to 0 = little/no linear relationship

Activity correlation findings: Daily steps were strongly positively correlated with total active hours (r = 0.77) and moderately positively correlated with calories burned (r = 0.58). Sedentary hours showed a moderate negative relationship with total active hours (r = -0.52), suggesting that more sedentary days tend to include less active time.

## MERGE ACTIVITY + SLEEP 
First, make the sleep date column match the activity date column name:

In [247]:
sleep_merge = sleep_clean.rename(columns={'SleepDay': 'ActivityDate'})

In [248]:
activity_sleep = pd.merge(
    activity_all,
    sleep_merge,
    on=['Id', 'ActivityDate'],
    how='inner'
)

## INNER MERGE
Inner used because we only want rows where:
the user exists in both datasets
and the date exists in both datasets

## INSPECT MERGED DATA

In [249]:
activity_sleep.shape

(410, 35)

In [250]:
activity_sleep.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,...,TotalMinutesAsleep,TotalTimeInBed,SleepDay_UK,HoursAsleep,HoursInBed,TimeAwakeInBedHours,SleepDuration,TimeInBed,AwakeInBedTime,DayOfWeek_y
0,1503960366,2016-04-12,13162,8.50,8.50,0.0,1.88,0.55,6.06,0.0,...,327,346,12/04/2016,5.450000,5.766667,0.316667,5h 27m,5h 46m,0h 19m,Tuesday
1,1503960366,2016-04-13,10735,6.97,6.97,0.0,1.57,0.69,4.71,0.0,...,384,407,13/04/2016,6.400000,6.783333,0.383333,6h 24m,6h 47m,0h 23m,Wednesday
2,1503960366,2016-04-15,9762,6.28,6.28,0.0,2.14,1.26,2.83,0.0,...,412,442,15/04/2016,6.866667,7.366667,0.500000,6h 52m,7h 22m,0h 30m,Friday
3,1503960366,2016-04-16,12669,8.16,8.16,0.0,2.71,0.41,5.04,0.0,...,340,367,16/04/2016,5.666667,6.116667,0.450000,5h 40m,6h 7m,0h 27m,Saturday
4,1503960366,2016-04-17,9705,6.48,6.48,0.0,3.19,0.78,2.51,0.0,...,700,712,17/04/2016,11.666667,11.866667,0.200000,11h 40m,11h 52m,0h 12m,Sunday


In [251]:
(activity_sleep['DayOfWeek_x'] == activity_sleep['DayOfWeek_y']).all()

np.True_

In [252]:
'ActivityDate_UK' in activity_sleep.columns

True

## MERGED DATA
Pandas have kept only rows where both datasets have the same ID and the same Date.
Merged shape suggests every cleaned sleep record successfully found a corresponding activity record since the cleaned sleep dataset had 410 rows also. 
DayOfWeek_y -> checked to see if that was identiacl to a DayOfWeek_x
ActvityDate_UK wasn't shown so checked to see if it was present

## TIDY MERGED DATA FRAME

In [253]:
activity_sleep = activity_sleep.drop(columns=['DayOfWeek_y'])

activity_sleep = activity_sleep.rename(
    columns={'DayOfWeek_x' : 'DayOfWeek'}
)

## SLEEP/ACTIVITY RELATIONSHIP ANALYSIS
Using correlation to see whether things like more steps or more active time are associated with more sleep.

In [254]:
sleep_activity_corr = activity_sleep[
    [
        'TotalSteps',
        'TotalActiveHours',
        'SedentaryHours',
        'Calories',
        'HoursAsleep',
        'HoursInBed',
        'TimeAwakeInBedHours'
    ]
].corr().round(2)

sleep_activity_corr

,TotalSteps,TotalActiveHours,SedentaryHours,Calories,HoursAsleep,HoursInBed,TimeAwakeInBedHours
TotalSteps,1.00,0.74,-0.13,0.41,-0.19,-0.17,0.03
TotalActiveHours,0.74,1.00,-0.26,0.39,-0.07,-0.10,-0.09
SedentaryHours,-0.13,-0.26,1.00,0.10,-0.60,-0.62,-0.17
Calories,0.41,0.39,0.10,1.00,-0.03,-0.13,-0.29
HoursAsleep,-0.19,-0.07,-0.60,-0.03,1.00,0.93,-0.00
HoursInBed,-0.17,-0.10,-0.62,-0.13,0.93,1.00,0.37
TimeAwakeInBedHours,0.03,-0.09,-0.17,-0.29,-0.00,0.37,1.00


## CORRELATION FINDINGS
Total steps and total active hours showed little relationship with sleep duration (r = -0.19 and r = -0.07, respectively). This was somewhat unexpected, as greater daily activity might reasonably be expected to correspond with longer sleep. Sedentary hours showed a stronger negative relationship with sleep duration (r = -0.60), although this should be interpreted cautiously because both measures represent portions of a finite 24-hour day. These correlations describe associations only and do not demonstrate causation.

## USER-LEVEL ACTIVITY/SLEEP SUMMARY
What does an average user look like ?

In [255]:
user_activity_summary = activity_all.groupby('Id').agg(
    AvgSteps=('TotalSteps', 'mean'), 
    AvgActiveHours=('TotalActiveHours', 'mean'),
    AvgSedentaryHours=('SedentaryHours', 'mean'),
    AvgCalories=('Calories', 'mean'),
    DaysRecorded=('ActivityDate', 'count')
).round(2)

user_activity_summary.head()

,AvgSteps,AvgActiveHours,AvgSedentaryHours,AvgCalories,DaysRecorded
Id,,,,,
1503960366,12174.80,4.74,14.17,1844.63,49
1624580081,5137.41,2.53,21.32,1448.63,49
1644430081,7780.92,3.81,18.83,2837.58,40
1844505072,2944.50,2.20,19.60,1613.57,42
1927972279,1298.71,1.04,20.68,2225.31,42


In [256]:
user_sleep_summary = activity_sleep.groupby('Id').agg(
    AvgSteps=('TotalSteps', 'mean'),
    AvgActiveHours=('TotalActiveHours', 'mean'),
    AvgSedentaryHours=('SedentaryHours', 'mean'),
    AvgSleepHours=('HoursAsleep', 'mean'),
    SleepDays=('ActivityDate', 'count')
).round(2)

user_sleep_summary.head()

,AvgSteps,AvgActiveHours,AvgSedentaryHours,AvgSleepHours,SleepDays
Id,,,,,
1503960366,12405.68,4.86,12.65,6.00,25
1644430081,7967.75,4.39,15.34,4.90,4
1844505072,3477.00,2.46,7.39,10.87,3
1927972279,1490.00,1.42,16.29,6.95,5
2026352035,5618.68,4.28,10.90,8.44,28


## USER-LEVEL ACTIVITY/SLEEP FINDINGS:
User-level participation and behaviour varied considerably. Some participants contributed substantially more activity records than others, while average activity levels also differed between users. Sleep coverage was particularly uneven, with some users contributing only 3–5 sleep records while others contributed more than 25.

## INSPECT DISTRIBUTION
Inspect the distribution of DaysRecorded and SleepDays to assess how evenly participants are represented in the activity and sleep datasets.

In [257]:
user_activity_summary['DaysRecorded'].describe()

count    35.000000
mean     39.228571
std       9.490198
min       8.000000
25%      38.500000
50%      42.000000
75%      42.000000
max      62.000000
Name: DaysRecorded, dtype: float64

In [258]:
user_sleep_summary['SleepDays'].describe()

count    24.000000
mean     17.083333
std      11.378533
min       1.000000
25%       4.750000
50%      20.500000
75%      27.250000
max      31.000000
Name: SleepDays, dtype: float64

In [259]:
display(
    user_activity_summary
    .sort_values('DaysRecorded')
    .head()
)

display(
    user_activity_summary
    .sort_values('DaysRecorded', ascending=False)
    .head()
)

,AvgSteps,AvgActiveHours,AvgSedentaryHours,AvgCalories,DaysRecorded
Id,,,,,
2891001357,773.62,4.19,18.33,2273.38,8
6391747486,1336.89,0.67,21.03,1763.11,9
8253242879,5061.33,2.07,21.80,1717.60,30
3372868164,6616.93,5.47,18.00,1908.83,30
2347167796,9948.59,4.94,11.56,2084.47,32


,AvgSteps,AvgActiveHours,AvgSedentaryHours,AvgCalories,DaysRecorded
Id,,,,,
4020332650,4114.95,1.95,19.54,2773.00,62
1624580081,5137.41,2.53,21.32,1448.63,49
1503960366,12174.80,4.74,14.17,1844.63,49
4702921684,8553.00,4.57,12.30,2983.42,45
4445114986,4729.13,3.64,14.22,2192.09,45


In [260]:
display(
    user_sleep_summary
    .sort_values('SleepDays')
    .head()
)

display(
    user_sleep_summary
    .sort_values('SleepDays', ascending=False)
    .head()
)

,AvgSteps,AvgActiveHours,AvgSedentaryHours,AvgSleepHours,SleepDays
Id,,,,,
2320127002,5079.00,4.03,18.82,1.02,1
7007744171,5115.50,3.67,19.14,1.14,2
6775888955,3499.00,1.78,16.07,5.83,3
1844505072,3477.00,2.46,7.39,10.87,3
8053475328,19078.67,5.02,13.96,4.95,3


,AvgSteps,AvgActiveHours,AvgSedentaryHours,AvgSleepHours,SleepDays
Id,,,,,
6962181067,9794.81,4.79,11.04,7.47,31
8378563200,8717.71,3.75,11.94,7.42,31
5553957443,8612.58,4.04,11.14,7.72,31
3977333714,11218.00,4.38,11.94,4.89,28
4445114986,4756.18,3.63,13.12,6.42,28


## DISTRIBUTION FINDINGS
Participation findings: The combined activity dataset contained 35 unique users, with 8–62 recorded activity days per user. Sleep coverage was substantially less complete, with only 24 users contributing between 1 and 31 records. Therefore, user-level sleep averages based on very few observations should be interpreted cautiously. The higher user count compared with the later one-month dataset (33 users) reflects the inclusion of participants appearing only in the earlier collection period.

## EXPORTING CSVs TO CREATE DATA VISAULS

In [261]:
activity_by_day.reset_index().to_csv(
    "../outputs/activity_by_day.csv",
    index=False
)

In [262]:
sleep_by_day.reset_index().to_csv(
    "../outputs/sleep_by_day.csv",
    index=False
)

In [263]:
activity_sleep.to_csv(
    "../outputs/activity_sleep.csv",
    index=False
)

In [264]:
coverage = pd.DataFrame({
    'Dataset': ['Activity', 'Sleep'],
    'UniqueUsers': [
        activity_all['Id'].nunique(),
        sleep_clean['Id'].nunique()
    ]
})

coverage

,Dataset,UniqueUsers
0,Activity,35
1,Sleep,24


In [265]:
coverage.to_csv(
    "../outputs/data_coverage.csv",
    index=False
)